pkg/monitortests/node/watchpods/cache_check.go

```diff
@@ -0,0 +1,81 @@
+package watchpods
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	coreinformers "k8s.io/client-go/informers/core/v1"
+	"k8s.io/client-go/kubernetes"
+)
+
+func checkCacheState(ctx context.Context, kubeClient kubernetes.Interface, podInformer coreinformers.PodInformer) ([]string, error) {
+	// We are running into problems where the cached list of pods contains pods that don't exist in the kube-apiserver.
+	// Let's try to check exactly.
+	cachedPods, err := podInformer.Lister().List(labels.Everything())
+	if err != nil {
+		return nil, fmt.Errorf("error listing cached pods: %w", err)
+	}
+
+	return checkCacheStateFromList(ctx, kubeClient, cachedPods)
+}
+
+func checkCacheStateFromList(ctx context.Context, kubeClient kubernetes.Interface, cachedPods []*corev1.Pod) ([]string, error) {
+	if len(cachedPods) == 0 {
+		return nil, fmt.Errorf("somehow the lister is missing pods")
+	}
+
+	biggestResourceVersion := -1
+	for _, cachedPod := range cachedPods {
+		if len(cachedPod.ResourceVersion) == 0 {
+			return nil, fmt.Errorf("encounted cached pod/%s -n %s without a resourceversion", cachedPod.Namespace, cachedPod.Name)
+		}
+		currRV, err := strconv.Atoi(cachedPod.ResourceVersion)
+		if err != nil {
+			return nil, fmt.Errorf("cached pod/%s -n %s has a non-integer RV %q: %w", cachedPod.Namespace, cachedPod.Name, cachedPod.ResourceVersion, err)
+		}
+		if currRV > biggestResourceVersion {
+			biggestResourceVersion = currRV
+		}
+	}
+
+	livePodsAtSameRV, err := kubeClient.CoreV1().Pods("").List(ctx, metav1.ListOptions{
+		ResourceVersion:      strconv.Itoa(biggestResourceVersion),
+		ResourceVersionMatch: metav1.ResourceVersionMatchExact,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("error listing pods resourceVersion=%v: %w", biggestResourceVersion, err)
+	}
+
+	failures := []string{}
+	for _, livePod := range livePodsAtSameRV.Items {
+		found := false
+		for _, cachedPod := range cachedPods {
+			if cachedPod.Namespace == livePod.Namespace && cachedPod.Name == livePod.Name {
+				found = true
+				if cachedPod.ResourceVersion != livePod.ResourceVersion {
+					failures = append(failures, "matching pod/%s -n %s has different resourceVersions: live=%v cached=%v", livePod.Name, livePod.Namespace, livePod.ResourceVersion, cachedPod.ResourceVersion)
+				}
+			}
+		}
+		if !found {
+			failures = append(failures, "live pod/%s -n %s RV=%v is not present in the cached results", livePod.Name, livePod.Namespace, livePod.ResourceVersion)
```

Comment:
in this [run](https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/29115/pull-ci-openshift-origin-master-e2e-aws-csi/1836559296096112640) the test failed because it didn't find a pod which must have been deleted from the cache and this wasn't reflected in the code that calculates the max RV